Mathlib/CategoryTheory/Sites/Sheafification.lean

```diff
@@ -0,0 +1,233 @@
+    change _ ≫ (sheafificationAdjunction J D).unit.app ((sheafToPresheaf J D).obj ⟨P, hP⟩) = _
+    erw [← inv_counit_map (sheafificationAdjunction J D) (X := ⟨P, hP⟩), comp_inv_eq_id]
+
+/-- If `P` is a sheaf, then `P` is isomorphic to `sheafify J P`. -/
+noncomputable def isoSheafify {P : Cᵒᵖ ⥤ D} (hP : Presheaf.IsSheaf J P) : P ≅ sheafify J P :=
+  letI := isIso_toSheafify J hP
+  asIso (toSheafify J P)
+
+@[simp]
+theorem isoSheafify_hom {P : Cᵒᵖ ⥤ D} (hP : Presheaf.IsSheaf J P) :
+    (isoSheafify J hP).hom = toSheafify J P :=
+  rfl
+
+/-- Given a sheaf `Q` and a morphism `P ⟶ Q`, construct a morphism from `sheafify J P` to `Q`. -/
+noncomputable def sheafifyLift {P Q : Cᵒᵖ ⥤ D} (η : P ⟶ Q) (hQ : Presheaf.IsSheaf J Q) :
+    sheafify J P ⟶ Q :=
+  (sheafificationAdjunction J D).homEquiv P ⟨Q, hQ⟩ |>.symm η |>.val
+
+@[reassoc (attr := simp)]
+theorem toSheafify_sheafifyLift {P Q : Cᵒᵖ ⥤ D} (η : P ⟶ Q) (hQ : Presheaf.IsSheaf J Q) :
+    toSheafify J P ≫ sheafifyLift J η hQ = η := by
+  rw [toSheafify, sheafifyLift, Adjunction.homEquiv_counit]
+  change _ ≫ (sheafToPresheaf J D).map _ ≫ _ = _
+  simp only [Adjunction.unit_naturality_assoc]
+  change _ ≫ (sheafificationAdjunction J D).unit.app ((sheafToPresheaf J D).obj ⟨Q, hQ⟩) ≫ _ = _
+  change _ ≫ _ ≫ (sheafToPresheaf J D).map _ = _
+  rw [sheafificationAdjunction J D |>.right_triangle_components (Y := ⟨Q, hQ⟩)]
+  simp
+
+theorem sheafifyLift_unique {P Q : Cᵒᵖ ⥤ D} (η : P ⟶ Q) (hQ : Presheaf.IsSheaf J Q)
+    (γ : sheafify J P ⟶ Q) : toSheafify J P ≫ γ = η → γ = sheafifyLift J η hQ := by
+  intro h
+  rw [toSheafify] at h
+  rw [sheafifyLift]
+  let γ' : (presheafToSheaf J D).obj P ⟶ ⟨Q, hQ⟩ := ⟨γ⟩
+  change γ'.val = _
+  rw [← Sheaf.Hom.ext_iff, ← Adjunction.homEquiv_apply_eq, Adjunction.homEquiv_unit]
+  exact h
+
+@[simp]
+theorem isoSheafify_inv {P : Cᵒᵖ ⥤ D} (hP : Presheaf.IsSheaf J P) :
+    (isoSheafify J hP).inv = sheafifyLift J (𝟙 _) hP := by
+  apply sheafifyLift_unique
+  simp [Iso.comp_inv_eq]
+
+theorem sheafify_hom_ext {P Q : Cᵒᵖ ⥤ D} (η γ : sheafify J P ⟶ Q) (hQ : Presheaf.IsSheaf J Q)
+    (h : toSheafify J P ≫ η = toSheafify J P ≫ γ) : η = γ := by
+  rw [sheafifyLift_unique J _ hQ _ h, ← h]
+  exact (sheafifyLift_unique J _ hQ _ h.symm).symm
+
+@[reassoc (attr := simp)]
+theorem sheafifyMap_sheafifyLift {P Q R : Cᵒᵖ ⥤ D} (η : P ⟶ Q) (γ : Q ⟶ R)
+    (hR : Presheaf.IsSheaf J R) :
+    sheafifyMap J η ≫ sheafifyLift J γ hR = sheafifyLift J (η ≫ γ) hR := by
+  apply sheafifyLift_unique
+  rw [← Category.assoc, ← toSheafify_naturality, Category.assoc, toSheafify_sheafifyLift]
+
+variable (D)
+
+@[simp]
+theorem sheafificationAdjunction_unit_app {P : Cᵒᵖ ⥤ D} :
+    (sheafificationAdjunction J D).unit.app P = toSheafify J P := rfl
+
+@[simp]
+theorem sheafificationAdjunction_counit_app_val (P : Sheaf J D) :
+    ((sheafificationAdjunction J D).counit.app P).val = sheafifyLift J (𝟙 P.val) P.cond := by
+  unfold sheafifyLift
+  rw [Adjunction.homEquiv_counit]
+  simp
```

Comment:
Can you please move these closer to where `sheafificationAdjunction` is defined?

Reply:
Moved them as high as I could